gap/ReducibleMatrixGroups.gi

```diff
@@ -223,22 +223,33 @@ function(d, q, k)
 
         # Now add a diagonal matrix where each of the SU(k, q) and SU(d - k, q)
         # blocks has determinant zeta ^ +- (q - 1).
-        determinantShiftMatrix := DiagonalMat(Concatenation(List([1..dHalf - 1], i -> 1),
+        # Note that the choice of matrix differs from the original Magma code,
+        # but this is much cleaner.
+        determinantShiftMatrix := DiagonalMat(Concatenation(List([1..dHalf - 1], i -> zeta ^ 0),
                                                             [zeta ^ (-1), zeta ^ (1 - q), zeta ^ q],
-                                                            List([dHalf + 3..d], i -> 1)));
+                                                            List([dHalf + 3..d], i -> zeta ^ 0)));
         Add(generators, determinantShiftMatrix);
     else
         # Find alpha, beta with alpha + alpha ^ q = 1 and beta * beta ^ q = -1.
-        alpha := SolveFrobeniusEquation("S", 1, q);
+        alpha := SolveFrobeniusEquation("S", zeta ^ 0, q);
         if IsOddInt(q) then
             beta := zeta ^ QuoInt(q - 1, 2);
         else
             beta := zeta ^ 0;
         fi;
         # We stabilise the subspace < e_1, ..., e_{kHalf}, w_1, f_{kHalf}, ..., f_1 >  
-        # and its complement < e_{kHalf + 1}, ..., e_{dHalf - 1}, w_2, f_{dHalf - 1}, ..., f_{kHalf + 1} >
+        # and its complement < e_{kHalf + 1}, ..., e_{dHalf - 1}, w_2, f_{dHalf - 1}, ..., f_{kHalf + 1} >,
+        # where w_1 = alpha * e_{dHalf} + f_{dHalf} and 
+        #       w_2 = -alpha ^ q * beta * e_{dHalf} + beta * f_{dHalf}.
         # (e_1, ..., e_{d / 2}, f_{d / 2}, ..., f_1 is the standard basis).
         #
+        # Note that, by construction of alpha and beta,
+        #       e_{dHalf} = w_1 + beta ^ q * w_2
+        #       f_{dHalf} = alpha ^ q * w_1 - alpha * beta ^ q * w_2
+        # Also by construction of alpha and beta, our standard unitary form
+        # applied to w_1 and w_1 or w_2 and w_2, respectively, each gives 1
+        # again, as needed.
+        #
         # The following matrices generate SU(k, q) x SU(d - k, q).
```

Comment:
Reading the code diagonally, it seems that it takes a matrix; chops it into blocks; then uses the blocks (plus some additional sprinkles) to assemble a new matrix.

Perhaps using `BlockMatrix` would allow to make this much more readable?

Reply:
I did think of that, but according to the GAP Manual, `BlockMatrix` requires all blocks to have the same dimensions - which isn't the case here.

Reply:
Ah indeed, it's been a long time I used it, I didn't realize it had this severe limitation. To bad. IMHO this kind of code looks a lot nicer in Julia, it has decent support for block matrices :-)